### PR TITLE
feat: Syntax parsing through AST

### LINKED ISSUE #2435: [Feature] Parse bindings with AST
## Summary
Replace current string based lookup for references with AST

## Motivation
We find dependencies to a particular binding with very coarse string based lookup today. This can be improved with AST for more robust parsing

### PR DESCRIPTION
**Pull request by eco-monk**
## Description

Syntax parsing through AST.
This PR is a follow up of #8617
(added comments on the changes I've made)

Fixes #2435

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- unit testing

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/ast-cyclic-dependency 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.16 **(0.08)** | 37.21 **(0.06)** | 34 **(0.13)** | 55.67 **(0.08)**
 :red_circle: | app/client/src/sagas/PostEvaluationSagas.ts | 16.89 **(-0.23)** | 0 **(0)** | 0 **(0)** | 20.49 **(-0.34)**
 :green_circle: | app/client/src/utils/DynamicBindingUtils.ts | 85.71 **(0.08)** | 70.48 **(0)** | 78.57 **(0)** | 85.26 **(0.1)**
 :red_circle: | app/client/src/workers/DataTreeEvaluator.ts | 77.16 **(-0.56)** | 61.7 **(0.08)** | 75.86 **(0)** | 76.87 **(-0.64)**
 :sparkles: :new: | **app/client/src/workers/ast.ts** | **92.78** | **78.57** | **90.91** | **94.51**
 :sparkles: :new: | **app/client/src/workers/constants.ts** | **100** | **100** | **100** | **100**
 :green_circle: | app/client/src/workers/evaluate.ts | 86.59 **(0.51)** | 70.83 **(0)** | 76.47 **(1.47)** | 86.25 **(0.54)**</details>

### CODE REVIEW

File: app/client/src/workers/ast.ts [hunk 1]
```diff
@@ -1,11 +1,248 @@
...
+  const variableDeclarations = new Set<string>();
+  // List of functionalParams found. This will be removed from the identifier list
+  const functionalParams = new Set<string>();
+  let ast: Node = { end: 0, start: 0, type: "" };
+  try {
+    ast = getAST(code);
+  } catch (e) {
+    if (e instanceof SyntaxError) {
+      // Syntax error. Ignore and return 0 identifiers
+      return [];
```
Comment (eco-monk) [inline app/client/src/workers/ast.ts:137]: Need to throw syntax errors in console.
DependencyMap/EvaluationOrder won't get updated incase of syntax errors. So any dependent entities will not be updated onChange.

Also noticed, some expressions get evaluated despite having syntax errors. e.g. IIFE without () - `function() {}()`
DependencyMap doesn't get updated either way.

Comment (eco-monk) [inline app/client/src/workers/ast.ts:137]: Need to throw syntax errors.
DependencyMap/EvaluationOrder won't get updated incase of syntax errors. So any dependent entities will not be updated onChange.

P.S. Also noticed, some expressions get evaluated successfully despite having syntax errors. e.g. IIFE without () - `function() {}()`
DependencyMap doesn't get updated either way.

File: app/client/src/workers/ast.ts [hunk 2]
```diff
@@ -1,11 +1,301 @@
...
+    FunctionDeclaration(node: Node) {
+      // params in function declarations are also counted as identifiers so we keep
+      // track of them and remove them from the final list of identifiers
+      if (!isFunctionDeclaration(node)) return;
+      functionalParams = new Set([
+        ...functionalParams,
+        ...getFunctionalParamsFromNode(node),
+      ]);
+    },
+    FunctionExpression(node: Node) {
```
Comment (eco-monk) [inline app/client/src/workers/ast.ts:243]: Filtering params from `FunctionExpresssion`

File: app/client/src/workers/ast.ts [hunk 3]
```diff
@@ -1,11 +1,301 @@
...
+       Some syntaxes won't be valid unless they're at the RHS of a statement.
+       Since we're assigning all code/script to RHS during evaluation, we do the same here.
+       So that during ast parse, those errors are neglected.
+    */
+    /* e.g. IIFE without braces
+      function() { return 123; }() -> is invalid
+      let result = function() { return 123; }() -> is valid 
+    */
+    const wrappedCode = wrapCode(unEscapedCode);
+    ast = getAST(wrappedCode);
```
Comment (eco-monk) [inline app/client/src/workers/ast.ts:167]: unescape code, wrap code with a function before parsing.

File: app/client/src/workers/ast.ts (hunk 1 above)
Comment (eco-monk) [reply app/client/src/workers/ast.ts:171]: Figured the issue. During evaluation, we wrap the code with a function before execution.
Hence, AST was throwing errors which don't really happen during evaluation.

Details added at `extractIdentifiersFromCode` method. Before `getAST` is called.

**Comment by ramsaptami** [comment]
Tested ok on deploy preview

**Review by hetunandu — APPROVED** [review]

**Pull request merged**